### DISCUSSION TITLE
The LSM-tree storage engine into OceanBase

### DISCUSSION THREAD

**Discussion by He-XiaoLong** [oceanbase/oceanbase#66] [Classroom] [answered]
![image](https://user-images.githubusercontent.com/57213420/120593424-10acad80-c472-11eb-82ea-e40334d1de09.png)
请问OB现使用的LSM-tree存储引擎只有两层吗？这种做法系统内的写放大是不是比较严重，相对于传统的多层来说

**Comment by simonjoylet** [comment, marked as answer]
There are four levels in lsm tree of ob 3.1. These levels are L0 memtable, L1 dump memtable, L2 Minor sstable, L3 major sstable.
LSM tree allow us to balance read amplification, write amplification, and space amplification by adjust the levels and compaction strategy.
OceanBase choose one strategy which prefers to reduce read amplification and space amplification by lower levels and reduce write amplification by reuse data blocks.

**Comment by He-XiaoLong** [comment]
Since most of the key ranges of sstable in the  L1  overlap, and there are four levels,  it is not very friendly to range queries. Does oceanbase do any optimization for range query (e.g., prefix or range bloom filter)?

**Comment by simonjoylet** [comment]
Four levels mean range scan in each level and a four-way merge. Only data from L1 may overlap. OceanBase try to keep a small number of sstable in L1. 
There is bloom filter for point query in OceanBase. For range query, since OceanBase support boom filter with prefix rowkey columns, if your query has prefix columns which match the columns in bloom filter, then the bloom filter will work. For other algorithms of range bloom filter, pr is welcome.

**Comment by simonjoylet** [comment]
OceanBase build bloom filter automatically by monitoring empty reads. If the count of empty read is more than threshold of bf_cache_miss_count_threshold, the bloom filter will be built as an async task. If the bloom filter build task is triggered by point query, the columns of bloom filter are all rowkey columns. If the bloom filter build task is triggered by range query, the columns of bloom filter is the common prefix columns of the query.
See function schedule_build_bloomfilter in "src/storage/ob_partition_scheduler.cpp"

**Reply by He-XiaoLong** [reply]
Hi ！From your Q&A community, I know that your memory only stores changed data (i.e., some columns of a row of data). In this way, Is it possible for all of a row of data to be distributed on different levels of LSM-tree? It seems that the cost of obtaining a whole row of data is relatively high.

**Reply by simonjoylet** [reply]
yeah, the columns of a row may be distributed in different levels of LSM-tree. However, OceanBase stores a compact row as the first row in sstables. The compact rows contains all columns of this row of latest version. In most cases, users read the latest row, so we just output the compact row in each sstable.

**Reply by He-XiaoLong** [reply]
sorry, i don't understand that "a compact row as the first row in sstables" .  can you give me an  example?
